app/src/main/java/io/ushakov/bike_workouts/data_engine/WorkoutDataProcessor.kt

```diff
@@ -150,57 +151,158 @@ class WorkoutDataProcessor(
+            val durationId = durationRepository.insert(activeDuration!!)
+            activeDuration!!.id = durationId
+            //createNewDuration()
+            //Log.d("DBG", "Duration created with $durationId id")
+            //Log.d("DBG", "Workout created with $workoutId id")
         }
     }
 
     fun pauseWorkout() {
         val workout = activeWorkout ?: return
+        val onGoingDuration = activeDuration ?: return
+
 
-        calculateCalories(Date())
 
         coroutineScope.launch(Dispatchers.IO) {
             workoutRepository.setWorkoutStatus(workout.id, false)
 
 
+            //TODO remove it after saving into state or observable whatever.
+            //get active duration from DB
+            /*val getDurationJob = async {
+                durationRepository.getById(activeWorkout!!.id)
+            }*/
+
+            //update it
+            //TODO Use following line after saving durationToUpdate into state
+            //activeDuration?.stopAt = Date()
+            //val durationToUpdate = getDurationJob.await()
+            //durationToUpdate.stopAt = Date()
+            onGoingDuration.stopAt = Date()
+            activeDuration!!.stopAt = Date()
+            //TODO remove async after feature is tested
+            val updateDurationJob = async { durationRepository.update(onGoingDuration) }
+            if ( updateDurationJob.await() > 0) {
+                //Log.d("DBG", "Duration updated at pause")
+            } else {
+                //Log.d("DBG", "Failed to update Duration")
+            }
+            //calculate it after stop time is written
+            //TODO fetch all duration
+            // calculate time duration of all
+            // pass that time duration to calculateCalories
+            totalWorkoutDuration = getWorkoutTotalDuration()
+            //Log.d("DBG", "Total workout duration when paused: $totalWorkoutDuration")
+
+            calculateCalories(totalWorkoutDuration)
+
             Log.d("DBG", "Workout paused")
         }
     }
 
     fun resumeWorkout() {
         val workout = activeWorkout ?: return
+        ///var newDuration = activeDuration ?: return
 
         coroutineScope.launch(Dispatchers.IO) {
             workoutRepository.setWorkoutStatus(workout.id, true)
 
-            Log.d("DBG", "Workout resumed")
+            //TODO create separate suspend function
+            val newDuration = Duration(0, workout.id, Date(), null)
+            val durationId = durationRepository.insert(newDuration)
+            newDuration.id = durationId
+            activeDuration = newDuration
+            //Log.d("DBG", "Duration created for resume workout with $durationId id")
+            //Log.d("DBG", "Workout resumed")
         }
     }
 
     fun stopWorkout() {
         val workout = activeWorkout ?: return
+        val onGoingDuration = activeDuration ?: return
+
+        CoroutineScope(Dispatchers.IO).launch {
```

Comment:
```suggestion
        coroutineScrope.launch(Dispatchers.IO) {
```
coroutineScope passed into WorkoutDataProcessor should be used instead